### DISCUSSION TITLE
[Layout] Primitive layout components

### DISCUSSION THREAD

**Discussion by sarahill** [Shopify/polaris-react#7195] [General]
As part of the [Layout foundations work](https://vault.shopify.io/gsd/projects/30116), the team is working on prototyping several primitive layout components based on the [Braid design system](https://seek-oss.github.io/braid-design-system/foundations/layout/). 

The goal of this discussion is to have a place to talk about the various layout primitives, how we want to approach them, and their job/purpose for building various layouts in the admin.

Primitive layout components for discussion:

- Box
- Card
- Stack
- Inline
- Columns
- Tiles
- ContentBlock
- Bleed

Related links:
- [Card discussion](https://github.com/Shopify/polaris/discussions/6458)
- [Modal flexibility](https://github.com/Shopify/polaris/discussions/6455)

**Comment by sarahill** [comment]
## `Tiles` or `Grid`

In our sync the other day the discussion of `Tile` came up. It wasn't clear what the purpose of the `Tile` component was and/or if another layout primitive could do the same job.

Here's my general understanding of what `Tile` does:
- The elements would wrap within the container to create a grid of elements
- The elements wrapped in `Tile` equal height and width
- `Tile` would be responsive -> you could opt to make the layout single column or it would be single column be default on mobile size screens

Here are some examples from the admin that I think fit that description (although not sure about the second example since that spans a couple rows)

<img width="1627" alt="image" src="https://user-images.githubusercontent.com/8629173/190512207-5990ebe1-5852-426e-a0d7-64335ef69820.png">

Please challenge any of that and bring up any additional thoughts and questions.

cc: @chazdean

**Reply by lemonmade** [reply]
In checkout, we have a `Tiles` component that we still provide for post-purchase extensions: https://shopify.dev/api/checkout-extensions/components/tiles. However, we intend to deprecate it there, and we have not brought it over to checkout UI extensions. In its place, we provide a `Grid` component where you can provide columns sizes, including potentially different numbers of columns on different conditions:

```tsx
function MyLayout() {
  return (
    <Grid
      columns={Style.default(['fill', 'fill']).when(
        {viewportInlineSize: {min: 'medium'}},
        ['fill', 'fill', 'fill'],
      )}
    >
      <View>Child 1</View>
      <View>Child 2</View>
      <View>Child 3</View>
      <View>Child 4</View>
    </Grid>
  );
}
```

You can also set the row size, although this does not force the content to be the same size by default. I don't think we have a way of forcing this right now, although I will let @jffortier confirm. We also have a `GridItem` component, https://shopify.dev/api/checkout-extensions/checkout/components/griditem, which allow developers to implement custom row and column spans (e.g., to achieve the second layout in the image you've included).

I believe our rationale for this change was that the grid component was needed anyways, and that the difference in usability of `Tiles` versus this `Grid` API was not enough to warrant the duplication of having both.

**Reply by chazdean** [reply]
Thank you for sharing those examples @sarahill, and I think it's safe to say we have the same understanding of what `Tile` would/should be used for currently.

I think what I struggle with is the point @lemonmade brought up which is, we also already have a [Grid](https://polaris.shopify.com/components/grid#props) component and when given the right props can act exactly like the current `Tile` prototype with more customization.

Not to say `Tile` can't do what `Grid` can, it just may require the use of other components like `Inline` to get more control of rows but I agree that it's hard to warrant having both right now.

**Reply by sarahill** [reply]
> I think what I struggle with is the point @lemonmade brought up which is, we also already have a [Grid](https://polaris.shopify.com/components/grid#props) component and when given the right props can act exactly like the current Tile prototype with more customization.
>
> Not to say Tile can't do what Grid can, it just may require the use of other components like Inline to get more control of rows but I agree that it's hard to warrant having both right now.

I think this is what I'm trying to understand. `Grid` can do everything but do we want it to or is the point that we want to break it down into those smaller pieces.

Some more questions/thoughts I'd like to know your perspective on:
- What are the trade-offs to these two approaches?
- The `Grid` component feels highly flexible but with that level of flexibility does it introduce a larger learning curve? (might just be a documentation problem)
- `Tiles` feels like it may support that idea of composability more to your point, Chaz, of using the other layout primitives with it to achieve various layouts.

**Reply by aveline** [reply]
Maybe a good next step would be to take one of those layouts that @sarahill shared and recreate it with both `Tile` and `Grid` and see how it feels. That might provide us with answers to some of those points you brought up @chazdean.

**Reply by heyjoethomas** [reply]
This is a great point. Especially if both are using css grid as the underlying technology. At that point it becomes redundent. But if not, then I think we would need both as css grid can do a lot more like a masonary structure that wouldn't fit the profile of a `Tile` component.

**Reply by chazdean** [reply]
> Maybe a good next step would be to take one of those layouts that @sarahill shared and recreate it with both `Tile` and `Grid` and see how it feels.

I agree! I went ahead and created an [issue](https://github.com/orgs/Shopify/projects/2250/views/59) just for this and will report back with my findings 👍🏾

**Reply by jffortier** [reply]
Edit: I realized I didnt threaded my comment, so I deleted the previous and copied it here.

---

> You can also set the row size, although this does not force the content to be the same size by default. I don't think we have a way of forcing this right now, although I will let @jffortier confirm.

The Checkout's Grid component with stretch the items to fill their cell. If no sizes are set for each row, the row will follow the tallest item.

This Grid will create 3 equal columns and 2 equal rows (250px).
```tsx
<Grid columns={['fill', 'fill', 'fill']} minBlockSize={500}>
    <View border="base" inlineAlignment="center" blockAlignment="center">content</View>
    <View border="base" inlineAlignment="center" blockAlignment="center">content</View>
    <View border="base" inlineAlignment="center" blockAlignment="center">content</View>
    <View border="base" inlineAlignment="center" blockAlignment="center">content</View>
    <View border="base" inlineAlignment="center" blockAlignment="center">content</View>
    <View border="base" inlineAlignment="center" blockAlignment="center">content</View>
</Grid>
```

![image](https://user-images.githubusercontent.com/739905/190661472-1fa3a109-f8da-4cd2-96e5-3f4126fac863.png)

**Reply by chazdean** [reply]
Reporting back my findings from my previous post. 

The main takeaways are:

- When it comes to the most simplistic grid layouts as shown in Example 1 and the 3rd layout option above, `Tile` is most likely a better pick as it it provides a more simplistic approach as it doesn't require its `children` wrapped. `Grid` wraps its `children` because it's more geared towards handling complex layouts that require each row needing utility.
- Example 2 is interesting because although `Tile` naturally doesn't default to support row customization, due to the fact its a primitive component it can be used in conjunction with other primitive components (`AlphaStack` and `Inline`) to complete the layout. `Grid` can also accomplish this task without the assistance of other components, so here it comes down to preference.

I think it really comes down to what we want our design approach to be as a Team, although they do overlap in some regards they do different things better. 

Below are some comparisons between `Tile` and `Grid` creating the same layout:

### Example 1 (Tile)
<kbd>
<img width="616" alt="Screen Shot 2022-09-21 at 2 10 14 PM" src="https://user-images.githubusercontent.com/59836805/191580413-4360f27c-0576-4e2f-9262-c55cbc2cbe41.png">
</kbd>

#### Code Snippet
```tsx
<Modal.Section>
  <Tile
    columns={{xs: 1, sm: 3, md: 4, lg: 4, xl: 4}}
    gap={{xs: '2', sm: '2', md: '2', lg: '2', xl: '2'}}
  >
    {placeHolderOne}
    {placeHolderOne}
    {placeHolderOne}
    {placeHolderOne}
    {placeHolderOne}
  </Tile>
</Modal.Section>
```
### Example 1 (Grid)
<kbd>
<img width="616" alt="Screen Shot 2022-09-21 at 2 10 34 PM" src="https://user-images.githubusercontent.com/59836805/191581612-489cdee1-7254-4f97-91c8-9a1fb6169661.png">
</kbd>

#### Code Snippet
```tsx
<Modal.Section>
  <Grid
    columns={{xs: 1, sm: 3, md: 4, lg: 4, xl: 4}}
    gap={{
      xs: 'var(--p-space-3)',
      sm: 'var(--p-space-3)',
      md: 'var(--p-space-3)',
      lg: 'var(--p-space-3)',
      xl: 'var(--p-space-3)',
    }}
  >
    <Grid.Cell>{placeHolderOne}</Grid.Cell>
    <Grid.Cell>{placeHolderOne}</Grid.Cell>
    <Grid.Cell>{placeHolderOne}</Grid.Cell>
    <Grid.Cell>{placeHolderOne}</Grid.Cell>
    <Grid.Cell>{placeHolderOne}</Grid.Cell>
  </Grid>
</Modal.Section>
```
### Example 2 (Tile)
<kbd>
<img width="640" alt="Screen Shot 2022-09-21 at 2 11 02 PM" src="https://user-images.githubusercontent.com/59836805/191582410-6935c7de-d204-436f-924c-9057cee434bc.png">
</kbd>

#### Code Snippet
```tsx
<Card title="Media (Tile)" sectioned>
  <AlphaStack spacing="2">
    <Inline spacing="3">
      {placeHolderTwo}
      <Tile
        columns={{xs: 2, sm: 2, md: 2, lg: 2, xl: 2}}
        gap={{xs: '3', sm: '3', md: '3', lg: '3', xl: '3'}}
      >
        {placeHolderOne}
        {placeHolderOne}
        {placeHolderOne}
        {placeHolderOne}
      </Tile>
    </Inline>
    <Tile
      columns={{xs: 4, sm: 4, md: 4, lg: 4, xl: 4}}
      gap={{xs: '3', sm: '3', md: '3', lg: '3', xl: '3'}}
    >
      {placeHolderOne}
      {placeHolderOne}
      {placeHolderOne}
      {placeHolderOne}
      {placeHolderOne}
    </Tile>
  </AlphaStack>
</Card>
```
### Example 2 (Grid)
<kbd>
<img width="653" alt="Screen Shot 2022-09-21 at 2 11 36 PM" src="https://user-images.githubusercontent.com/59836805/191582685-e48b3e20-dfe7-44b4-bb9c-65781de09e5b.png">
</kbd>

#### Code Snippet
```tsx
<Card title="Media (Grid)" sectioned>
  <Grid
    columns={{xs: 4, sm: 4, md: 4, lg: 4, xl: 4}}
    gap={{
      xs: 'var(--p-space-3)',
      sm: 'var(--p-space-3)',
      md: 'var(--p-space-3)',
      lg: 'var(--p-space-3)',
      xl: 'var(--p-space-3)',
    }}
    areas={{
      xs: ['a a b c', 'a a d e', 'f g h i', 'j k . .'],
      sm: ['a a b c', 'a a d e', 'f g h i', 'j k . .'],
      md: ['a a b c', 'a a d e', 'f g h i', 'j k . .'],
      lg: ['a a b c', 'a a d e', 'f g h i', 'j k . .'],
      xl: ['a a b c', 'a a d e', 'f g h i', 'j k . .'],
    }}
  >
    <Grid.Cell area="a">{placeHolderTwo}</Grid.Cell>
    <Grid.Cell area="b">{placeHolderOne}</Grid.Cell>
    <Grid.Cell area="c">{placeHolderOne}</Grid.Cell>
    <Grid.Cell area="d">{placeHolderOne}</Grid.Cell>
    <Grid.Cell area="e">{placeHolderOne}</Grid.Cell>
    <Grid.Cell area="f">{placeHolderOne}</Grid.Cell>
    <Grid.Cell area="g">{placeHolderOne}</Grid.Cell>
    <Grid.Cell area="h">{placeHolderOne}</Grid.Cell>
    <Grid.Cell area="i">{placeHolderOne}</Grid.Cell>
    <Grid.Cell area="j">{placeHolderOne}</Grid.Cell>
  </Grid>
</Card>
```

**Reply by sarahill** [reply]
Thanks for doing these explorations! Cool to see the examples built in different ways.

I think using `Tile` along with other primitives, like `Inline`, makes sense with our current direction. Each smaller component does one job and you combine them to get more complex layouts.

**Reply by aveline** [reply]
I agree with Sara, `Tile` looks like it works well for these examples. I wonder what an example of a layout where `Grid` would be the better choice looks like? Would it be something with uneven row heights? I think that's one of the limitations of `Tile` right?

**Reply by sarahill** [reply]
Thoughts from @kyledurand 

> Some of my concerns here:
> 
> 1. Let's align on naming in both Figma and in code. What are designers most commonly referring to? IMO I don't like the name tiles, when I think of wall tiles or floor tiles they can come in all shapes and sizes like mosaics etc... we're all familiar with a grid, we had grid paper in school, designers use grid overlays, developers use css grid. It's familiar and represents both sides.
> 
> 2. Grid is currently only used in 3 places in web but in some complex ways. Here's an example of a configuration of props: [https://github.com/Shopify/web/blob/main/app/sections/Companies/CompanyDetails/components/CompanyLocationListItem/utilities/constants.ts](https://www.google.com/url?q=https://github.com/Shopify/web/blob/main/app/sections/Companies/CompanyDetails/components/CompanyLocationListItem/utilities/constants.ts&sa=D&source=docs&ust=1664299522790733&usg=AOvVaw2l97Y1VIOmToi_hJXo8mFr)
> 
> Let's make sure we can achieve the same things with Tiles if we deprecate Grid.
> 
> 3. Why are we calling it Tile when the component represents a groups of Tiles? Braid calls it Tiles, let's call it Tiles if that's what we go with

**Reply by heyjoethomas** [reply]
From a designer pov I think `Tile` is just fine. I would have no issue with that name. Something like `Grid` has way trouble I think as designers will think mainly of a 12 column grid.

**Reply by sarahill** [reply]
When I think of `Tiles` I think of something pretty specific. When I think of `Grid` I tend to think about larger complex layouts or a traditional 12 column grid (as Joe mentioned).

If we had `Grid` and `Columns` that feels confusing as to know when I would use one over the other. Whereas with `Tiles` I do think of something different. 

It might feel a little new to designers as a term but I think with guidance and visuals it should be fine. I think designers have a mental image of what a **tile** is and the component generally matches that model (for me at least).

**Reply by kyledurand** [reply]
> If we had Grid and Columns that feels confusing 

Love that thinking. They both do basically the same thing but Grid allows for a bit more control over its layout which adds complexity and I don't think that's necessarily what we want. Let's roll with `Tiles` and remove `Grid`.

**Comment by sarahill** [comment]
## Utility styles

There's been some convos cropping up around shared utility styles and naming conventions. Kicking a thread to have some convo around that.

### Some conversation points:

- What utility styles do we need?
- What naming convections do we want to use for these styles?

**Spacing utilities:**
- `space`
- `top`, `bottom`, `left`, `right`, `vertical`, `horizontal`
  - When would I reach for these over a `padding` or `margin`?
- `padding`, `paddingX`, `paddingY`, `paddingTop`, `paddingBottom`, `paddingLeft`, `paddingRight`
  - Our current approach is to generally only use padding to keep everything contained and composed: Components shouldn't concern themselves with spacing outside their bounds. This also avoids issues with collapsing margins.
    - Is this true for both vertical and horizontal spacing?
- I think we need `margin` for `Bleed` though as it applies negative margin
- How do we want to handle negative values?

**Position**
- `top`, `bottom`, `left`, `right`

**Widths:**
- `minWidth`, `maxWidth`, ?

**Display:**
- `block`, `flex`, `inline`, `inlineBlock`, `none`

**Border-radius**
- These change in some cases on desktop vs. mobile

**Flex**
- `justifyContent`, `alignItems`, `flexDirection`, `flexWrap`, `flexShrink`, `flexGrow`, ?

These are just to start the convo to strengthen our decisions and opinions. Please challenge any of these. These are largely based off of [Braid's utility classes](https://seek-oss.github.io/braid-design-system/css/atoms/) and [Style system](https://github.com/styled-system/styled-system/blob/master/docs/table.md#core)

Related links:
- [Alpha Bleed component PR](https://github.com/Shopify/polaris/pull/7156/)

**Comment by sarahill** [comment]
## Card component

As I'm working on adding `AlphaCard` to the UI Kit I'm starting to think it needs to be more opinionated.

Here are my thoughts / observations:
The first value in the "Values" column represents the default value with the other value options below. For things like `backgroundColor`, `borderRadius`, and `elevation` I think those should actually be pretty opinionated at this level. Each of those have 2 options. 

Anything that needs to go outside of these styles would use the `Box` component. 

If I'm interpreting it correctly, it looks like there's more flexibility in `AlphaCard` styling right now for `backgroundColor` and `borderRadius` correct me if I'm wrong though. I know we were still figuring out some of these opinions.

Thoughts?

| Props | Values |
|---|---|
| `children` | `React.ReactNode;` |
| `backgroundColor` | `surface` |
| | `surface-subdued` |
| `borderRadius` | `border-radius-2` |
| | `none` |
| `elevation` | `shadow-card` |
| | `shadow-transparent` |
| `padding` | `space-5 ` |
| | space tokens scale |
| `roundedAbove` | breakpoint tokens |

**Reply by sarahill** [reply]
I do wonder about `elevation` though. Thinking about some of @zecarlostorre visual design exploration with elevation and how that may change based on level of importance within the UI. I don't think that affects `Cards` as much though as other elements like a `popover` or maybe a `modal`. Thoughts @zecarlostorre?

**Reply by aveline** [reply]
Yes right now we are allowing the following for `borderRadius`

`BorderRadiusTokenScale = "base" | "05" | "1" | "2" | "3" | "4" | "5" | "6" | "large" | "half"`

and for `backgroundColor`

`CardBackgroundColorTokenScale = "surface" | "surface-dark" | "surface-neutral" | "surface-neutral-hovered" | "surface-neutral-pressed" | "surface-neutral-disabled" | "surface-neutral-subdued" | "surface-neutral-subdued-dark" | ... 32 more ... | "surface-attention"`

We can definitely tighten up the permitted background colors, as well as border radius. My only thought for border radius was I thought we wanted to allow more flexibility for the nested cards use case where border radius would need to be adjusted on children cards?

Perhaps we allow `2` and all the values below that? Or `"1" | "2"`  if we anticipate only nesting one level? Or like you said, using the `Box` component for that could be the workaround.

**Reply by sarahill** [reply]
> We can definitely tighten up the permitted background colors, as well as border radius. My only thought for border radius was I thought we wanted to allow more flexibility for the nested cards use case where border radius would need to be adjusted on children cards?

That's a good point re: nested cards. My thought here is that putting a `Card` within a `Card` without changing its styling doesn't feel great but I know there are instances of that happening today. My assumption is this is largely due to the fact that we don't have any way to style containers and there's no guidance on nesting containers and what they should look like.

I'm leaning towards starting with the `Card` a little more on the strict side and seeing if that works or breaks down. It feels better to give more flexibility instead of taking it away and since we've got `Box` it feels low risk. Feel free to challenge though.

@heyjoethomas do you have opinions here?

**Reply by aveline** [reply]
That approach sounds good to me @sarahill, it's much easier to extend the values allowed than remove them.

**Reply by heyjoethomas** [reply]
@sarahill Yeah I agree with this. Since `Card` is so fundamental to the design of the admin I think we need to be really careful introducing variation. I would lean more towards strictness because the impact of visual changes would be very high given almost everything in the admin is within a `Card`. Even something like border-radius I could see being locked down and then we just handle it at a certain breakpoint. 

> My only thought for border radius was I thought we wanted to allow more flexibility for the nested cards

I wouldn't think a `Card` can be nested in another `Card`. To me that would be a `Box` because it won't have the same background and specifically elevation of a `Card`.

For example, the grouped content inside the `Card` below wouldn't be a `Card` imo. It would use the `Box` component which would have much openness than the `Card` does. 

<img width="488" alt="Screen Shot 2022-09-27 at 9 38 14 AM" src="https://user-images.githubusercontent.com/5749440/192556843-202b0f7d-a66c-46f6-a9a3-5cb201c02acc.png">

**Reply by sarahill** [reply]
➕ to using `Box` for nested containers. There are examples of nested cards today but I think that's just guidance for migration and something to look at as part of visual design of containers in general.

<img width="996" alt="image" src="https://user-images.githubusercontent.com/8629173/192567597-b8555e80-ede2-4404-adc0-1a3c9a22eb83.png">

**Reply by heyjoethomas** [reply]
Totally agree. This feels very odd to me and doesn't seem like a design pattern we want to replicate.

**Reply by aveline** [reply]
In the example that you posted @heyjoethomas the nested content would be very easy to create with our existing `AlphaCard`. I'm not sure why a dev shouldn't use `AlphaCard` for that since it will provide the border, background color and flat elevation that already exists in `AlphaCard` and we enable the flexibility to create sections within.

**Reply by heyjoethomas** [reply]
I think a `Card` is a specific recipe of styles and is a larger wrapper than that sub-section. From a language perspective it also doesn't look or feel like a `Card`. It feels like a container or section of content. I think that mental model is important to keep. By making the subsection a `Card` we have made it pretty difficult to define what a `Card` is and what it isn't. I think keeping things clearly defined is important for long term health and scalability.

**Reply by sarahill** [reply]
☝️ I agree with that. If we're going to make `AlphaCard` really flexible then I would start to question why we have both Card and Box. It starts to get a little fuzzy as to when I should reach for one vs. the other.

**Reply by halfool-draws** [reply]
@sarahill - apologies for the delay, just managed to catch up with this.

Re:elevation - I think this approach makes sense right now, we'll figure out what we need when we land on a definite direction and then we see what things we may need to add, if anything.

Just to share what on my mind: I'm thinking beyond a shadow, we may also want to have a way to represent how bright or dim an object should be. But again, for now this is speculation, we don't need to get into that until we know exactly what we want to do.

**Comment by sarahill** [comment]
## `ContentBlock`

Let's clarify the intent of `ContentBlock`. There have been a few discussion in different places about the use case for this component in the admin so I wanted to consolidate those and align here.

My understanding is `ContentBlock` would set the max-width of the content within the current breakpoint. My thought was `Page` would be refactored to use `ContentBlock` to set the width of the main page content. For example:

<img width="1415" alt="image" src="https://user-images.githubusercontent.com/8629173/193646978-91f2db6f-19e4-4fb5-ae76-7c6d57cdf1f0.png">

So this would be a more opinionated layout component that specifically sets max-widths for the contents container.

Some questions to start the conversation:
- Other use cases for `ContentBlock`?
- Why not use `Box` for this instead of creating a separate component?

**Reply by aveline** [reply]
I think we should ship `ContentBlock` as is for the alpha version and try using it with `Page`. I'm hoping there will be some good learnings there that could answer some of those questions.

**Reply by sarahill** [reply]
I'm all for shipping it as is. My goal is for us to be able to be confident in why we need `ContentBlock` and what job it's doing. It sounds like we can learn and form stronger opinions as we start applying it to components 👍

**Reply by SeanyB** [reply]
So, in the context of Page, would that component continue to have props for narrowWidth, normalWidth, and fullWidth, but use ContentBlock to achieve them in the underlying implementation? Or would the user need to wrap their components in the ContentBlock and then put the ContentBlock inside the Page?

**Comment by aveline** [comment]
# Should `Box` have support for passing in an `id`?

Currently we do not have an `id` prop on `Box` to support passing in an `id` as an HTML attribute.

Are we concerned about the misuse of the `id` attribute and therefore not including it in our layout primitives? Should we be removing the use of `id` in our components?

**Reply by kyledurand** [reply]
It seems to me like it’s going to come down to preference. In theory we shouldn’t need to pass any ids if our library is perfectly composable.

[Braid allows ID on box](https://seek-oss.github.io/braid-design-system/playroom/#?code=N4Igxg9gJgpiBcIA8AhCAPABASygXgB0QAjDAWlyID5T0kB6NdKggOxABoQAXACxgC2MAM4IA2iACGAB0lgQAXQC%2BQA)

A couple of others don’t:
https://nulogy.design/components/box/
https://paste.twilio.design/primitives/box/

I’m leaning towards not including id to keep our APIs simple and clean but again that’s just preference

**Reply by aveline** [reply]
After further discussion we've decided to allow `id` on `Box` and `Text`. While we don't want to encourage misuse of the attribute, we recognize there are valid instances where using `id` is necessary and we don't want to force users to have to use workarounds like a wrapper div with `id`. Having it as a prop will make it easier to track and evaluate usage.
